Tests/NIOTests/ChannelTests.swift

```diff
@@ -2019,4 +2019,86 @@ public class ChannelTests: XCTestCase {
         try allDone.futureResult.wait()
         XCTAssertNoThrow(try sc.syncCloseAcceptingAlreadyClosed())
     }
+
+    func testSocketFailingAsyncCorrectlyTearsTheChannelDownAndDoesntCrash() throws {
+        // regression test for #302
+        enum DummyError: Error { case dummy }
+        class SocketFailingAsyncConnect: Socket {
+            private var firstReadHappened = false
+            init() throws {
+                try super.init(protocolFamily: PF_INET, type: Posix.SOCK_STREAM, setNonBlocking: true)
+            }
+            override func connect(to address: SocketAddress) throws -> Bool {
+                _ = try super.connect(to: address)
+                return false
+            }
+            override func finishConnect() throws {
+                throw DummyError.dummy
+            }
+        }
+        class VerifyThingsAreRightHandler: ChannelInboundHandler {
+            typealias InboundIn = Never
+            private let allDone: EventLoopPromise<Void>
+            enum State {
+                case fresh
+                case registered
+                case unregistered
+            }
+            private var state: State = .fresh
+
+            init(allDone: EventLoopPromise<Void>) {
+                self.allDone = allDone
+            }
+            deinit {
+                XCTAssertEqual(.unregistered, self.state)
+            }
+            func channelActive(ctx: ChannelHandlerContext) {
```

Comment:
nit: But an empty line between func`s so its easier to read ?

Reply:
@normanmaurer can do if you want, I removed them because I thought it's  soo many lines that are all boring. totally up to you though

Reply:
@normanmaurer proposed a new version with spaces